crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+        }
+        Ok(())
+    }
+
+    pub fn process_peer_request(&mut self, rpc_req: IncomingRpcRequest) -> Result<()> {
+        let IncomingRpcRequest {
+            msg,
+            mut response_sender,
+            sender,
+        } = rpc_req;
+        debug!(
+            "[JWK] process_peer_request: sender={}, is_self={}",
+            sender,
+            sender == self.my_addr
+        );
+        match msg {
+            JWKConsensusMsg::ObservationRequest(request) => {
+                let state = self.states_by_issuer.entry(request.issuer).or_default();
+                let response: Result<JWKConsensusMsg> = match &state.consensus_state {
+                    ConsensusState::NotStarted => Err(anyhow!("observed update unavailable")),
+                    ConsensusState::InProgress { my_proposal, .. }
+                    | ConsensusState::Finished { my_proposal, .. } => Ok(
+                        JWKConsensusMsg::ObservationResponse(ObservedUpdateResponse {
+                            epoch: self.epoch_state.epoch,
+                            update: my_proposal.clone(),
+                        }),
+                    ),
+                };
+                response_sender.send(response);
+                Ok(())
+            },
+            _ => {
+                bail!("unexpected rpc: {}", msg.name());
+            },
+        }
+    }
+
+    /// Triggered once the `certified_update_producer` produced a quorum-certified update.
+    pub fn process_quorum_certified_update(&mut self, update: QuorumCertifiedUpdate) -> Result<()> {
+        let issuer = update.update.issuer.clone();
+        let state = self.states_by_issuer.entry(issuer.clone()).or_default();
+        match &state.consensus_state {
+            ConsensusState::InProgress { my_proposal, .. } => {
+                //TODO: counters
+                let txn = ValidatorTransaction::ObservedJWKUpdate(update.clone());
+                let vtxn_guard =
+                    self.vtxn_pool
+                        .put(Topic::JWK_CONSENSUS(issuer), Arc::new(txn), None);
+                state.consensus_state = ConsensusState::Finished {
+                    vtxn_guard,
+                    my_proposal: my_proposal.clone(),
+                    quorum_certified: update.clone(),
+                };
+                info!("[JWK] qc update obtained, update={:?}", update);
+                Ok(())
+            },
+            _ => Err(anyhow!(
+                "qc update not expected for issuer {:?} in state {}",
+                update.update.issuer,
+                state.consensus_state.name()
+            )),
+        }
+    }
+}
+
+/// An instance of this resource is created when `JWKManager` starts the QC update building process for an issuer.
+/// Then `JWKManager` needs to hold it. Once this resource is dropped, the corresponding QC update process will be cancelled.
+#[derive(Clone, Debug)]
+pub struct QuorumCertProcessGuard {
+    handle: Option<AbortHandle>,
```

Comment:
This is optional only because of the test?

Reply:
fixed